File: frontend/lib/models/wrong_answer.dart
/**
 * 오답 노트 관련 데이터 모델
 * 
 * 사용자가 틀린 문제들을 관리하고 복습할 수 있도록 하는
 * 오답 노트 시스템의 데이터 모델을 정의합니다.
 */

class WrongAnswer {
  final int id;
  final int questionId;
  final String questionContent;
  final String userAnswer;
  final String correctAnswer;
  final String? explanation;
  final String? unit;
  final String? subject;
  final String? topic;
  final double difficulty;
  final int points;
  final String sourceType;
  final int reviewCount;
  final DateTime? lastReviewedAt;
  final bool mastered;
  final DateTime? masteredAt;
  final DateTime createdAt;

  const WrongAnswer({
    required this.id,
    required this.questionId,
    required this.questionContent,
    required this.userAnswer,
    required this.correctAnswer,
    this.explanation,
    this.unit,
    this.subject,
    this.topic,
    required this.difficulty,
    required this.points,
    required this.sourceType,
    required this.reviewCount,
    this.lastReviewedAt,
    required this.mastered,
    this.masteredAt,
    required this.createdAt,
  });

  factory WrongAnswer.fromJson(Map<String, dynamic> json) {
    return WrongAnswer(
      id: json['id'],
      questionId: json['question_id'],
      questionContent: json['question_content'],
      userAnswer: json['user_answer'],
      correctAnswer: json['correct_answer'],
      explanation: json['explanation'],
      unit: json['unit'],
      subject: json['subject'],
      topic: json['topic'],
      difficulty: (json['difficulty'] ?? 3.0).toDouble(),
      points: json['points'] ?? 3,
      sourceType: json['source_type'] ?? 'daily_test',
      reviewCount: json['review_count'] ?? 0,
      lastReviewedAt: json['last_reviewed_at'] != null
          ? DateTime.parse(json['last_reviewed_at'])
          : null,
      mastered: json['mastered'] ?? false,
      masteredAt: json['mastered_at'] != null
          ? DateTime.parse(json['mastered_at'])
          : null,
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'question_id': questionId,
      'question_content': questionContent,
      'user_answer': userAnswer,
      'correct_answer': correctAnswer,
      'explanation': explanation,
      'unit': unit,
      'subject': subject,
      'topic': topic,
      'difficulty': difficulty,
      'points': points,
      'source_type': sourceType,
      'review_count': reviewCount,
      'last_reviewed_at': lastReviewedAt?.toIso8601String(),
      'mastered': mastered,
      'mastered_at': masteredAt?.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
    };
  }

  /// 복습이 필요한지 확인
  bool get needsReview {
    if (mastered) return false;
    
    if (lastReviewedAt == null) return true;
    
    final daysSinceLastReview = DateTime.now().difference(lastReviewedAt!).inDays;
    return daysSinceLastReview >= 3;
  }

  /// 난이도 레벨 텍스트 반환
  String get difficultyText {
    if (difficulty <= 2.0) return '쉬움';
    if (difficulty <= 3.0) return '보통';
    if (difficulty <= 4.0) return '어려움';
    return '매우 어려움';
  }

  /// 출처 텍스트 반환
  String get sourceText {
    switch (sourceType) {
      case 'diagnostic':
        return '진단 평가';
      case 'daily_test':
        return '일일 모의고사';
      default:
        return '기타';
    }
  }
}

class WrongAnswerReview {
  final int id;
  final String userAnswer;
  final bool isCorrect;
  final int? timeTaken;
  final int? confidenceLevel;
  final String? notes;
  final DateTime createdAt;

  const WrongAnswerReview({
    required this.id,
    required this.userAnswer,
    required this.isCorrect,
    this.timeTaken,
    this.confidenceLevel,
    this.notes,
    required this.createdAt,
  });

  factory WrongAnswerReview.fromJson(Map<String, dynamic> json) {
    return WrongAnswerReview(
      id: json['id'],
      userAnswer: json['user_answer'],
      isCorrect: json['is_correct'],
      timeTaken: json['time_taken'],
      confidenceLevel: json['confidence_level'],
      notes: json['notes'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_answer': userAnswer,
      'is_correct': isCorrect,
      'time_taken': timeTaken,
      'confidence_level': confidenceLevel,
      'notes': notes,
    };
  }
}

class WrongAnswerStats {
  final int totalCount;
  final int masteredCount;
  final int reviewNeededCount;
  final Map<String, int> bySubject;
  final Map<String, int> byUnit;
  final Map<String, int> byDifficulty;
  final List<WrongAnswerReview> recentReviews;

  const WrongAnswerStats({
    required this.totalCount,
    required this.masteredCount,
    required this.reviewNeededCount,
    required this.bySubject,
    required this.byUnit,
    required this.byDifficulty,
    required this.recentReviews,
  });

  factory WrongAnswerStats.fromJson(Map<String, dynamic> json) {
    return WrongAnswerStats(
      totalCount: json['total_count'] ?? 0,
      masteredCount: json['mastered_count'] ?? 0,
      reviewNeededCount: json['review_needed_count'] ?? 0,
      bySubject: Map<String, int>.from(json['by_subject'] ?? {}),
      byUnit: Map<String, int>.from(json['by_unit'] ?? {}),
      byDifficulty: Map<String, int>.from(json['by_difficulty'] ?? {}),
      recentReviews: (json['recent_reviews'] as List<dynamic>? ?? [])
          .map((review) => WrongAnswerReview.fromJson(review))
          .toList(),
    );
  }

  /// 마스터율 계산
  double get masteryRate {
    if (totalCount == 0) return 0.0;
    return masteredCount / totalCount;
  }

  /// 복습 필요율 계산
  double get reviewNeededRate {
    if (totalCount == 0) return 0.0;
    return reviewNeededCount / totalCount;
  }
}

class WrongAnswerFilter {
  final String? subject;
  final String? unit;
  final double? difficultyMin;
  final double? difficultyMax;
  final bool? mastered;
  final String? sourceType;
  final int limit;
  final int offset;

  const WrongAnswerFilter({
    this.subject,
    this.unit,
    this.difficultyMin,
    this.difficultyMax,
    this.mastered,
    this.sourceType,
    this.limit = 20,
    this.offset = 0,
  });

  Map<String, dynamic> toQueryParams() {
    final params = <String, dynamic>{
      'limit': limit,
      'offset': offset,
    };

    if (subject != null) params['subject'] = subject;
    if (unit != null) params['unit'] = unit;
    if (difficultyMin != null) params['difficulty_min'] = difficultyMin;
    if (difficultyMax != null) params['difficulty_max'] = difficultyMax;
    if (mastered != null) params['mastered'] = mastered;
    if (sourceType != null) params['source_type'] = sourceType;

    return params;
  }
}

File: backend/models/wrong_answer.py
"""
오답 노트 관련 데이터베이스 모델

이 모듈은 사용자의 틀린 문제들을 관리하고 복습할 수 있도록 하는
오답 노트 시스템의 데이터 모델을 정의합니다.
"""

from sqlalchemy import Column, Integer, String, DateTime, Boolean, Text, ForeignKey, Float
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List
from .database import Base


class WrongAnswer(Base):
    """오답 노트 테이블
    
    사용자가 틀린 문제들을 저장하고 관리하는 테이블입니다.
    진단 평가와 일일 모의고사에서 틀린 문제들을 자동으로 수집합니다.
    """
    __tablename__ = "wrong_answers"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    question_id = Column(Integer, nullable=False)  # 문제 ID
    question_content = Column(Text, nullable=False)  # 문제 내용
    user_answer = Column(String(10), nullable=False)  # 사용자 답안
    correct_answer = Column(String(10), nullable=False)  # 정답
    explanation = Column(Text)  # 해설
    unit = Column(String(100))  # 단원 (예: "언어이해-인식론")
    subject = Column(String(50))  # 과목 (예: "언어이해")
    topic = Column(String(100))  # 주제 (예: "인식론")
    difficulty = Column(Float, default=3.0)  # 난이도 (1.0-5.0)
    points = Column(Integer, default=3)  # 배점
    source_type = Column(String(20), default="daily_test")  # 출처 (diagnostic, daily_test)
    source_id = Column(Integer)  # 출처 ID (진단평가 ID 또는 일일모의고사 ID)
    
    # 복습 관련 필드
    review_count = Column(Integer, default=0)  # 복습 횟수
    last_reviewed_at = Column(DateTime(timezone=True))  # 마지막 복습 시간
    mastered = Column(Boolean, default=False)  # 마스터 여부
    mastered_at = Column(DateTime(timezone=True))  # 마스터 달성 시간
    
    # 메타데이터
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # 관계 설정
    user = relationship("User", back_populates="wrong_answers")


class WrongAnswerReview(Base):
    """오답 노트 복습 기록 테이블
    
    사용자가 오답 노트의 문제를 다시 풀어본 기록을 저장합니다.
    """
    __tablename__ = "wrong_answer_reviews"
    
    id = Column(Integer, primary_key=True, index=True)
    wrong_answer_id = Column(Integer, ForeignKey("wrong_answers.id"), nullable=False)
    user_answer = Column(String(10), nullable=False)  # 복습 시 사용자 답안
    is_correct = Column(Boolean, nullable=False)  # 복습 시 정답 여부
    time_taken = Column(Integer)  # 소요 시간 (초)
    confidence_level = Column(Integer)  # 확신도 (1-5)
    notes = Column(Text)  # 사용자 메모
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # 관계 설정
    wrong_answer = relationship("WrongAnswer")


# Pydantic 모델들
class WrongAnswerCreate(BaseModel):
    """오답 노트 생성용 모델"""
    question_id: int
    question_content: str
    user_answer: str
    correct_answer: str
    explanation: Optional[str] = None
    unit: Optional[str] = None
    subject: Optional[str] = None
    topic: Optional[str] = None
    difficulty: Optional[float] = 3.0
    points: Optional[int] = 3
    source_type: str = "daily_test"
    source_id: Optional[int] = None


class WrongAnswerResponse(BaseModel):
    """오답 노트 응답용 모델"""
    id: int
    question_id: int
    question_content: str
    user_answer: str
    correct_answer: str
    explanation: Optional[str]
    unit: Optional[str]
    subject: Optional[str]
    topic: Optional[str]
    difficulty: Optional[float]
    points: Optional[int]
    source_type: str
    review_count: int
    last_reviewed_at: Optional[datetime]
    mastered: bool
    mastered_at: Optional[datetime]
    created_at: datetime
    
    class Config:
        from_attributes = True


class WrongAnswerReviewCreate(BaseModel):
    """오답 복습 기록 생성용 모델"""
    user_answer: str
    is_correct: bool
    time_taken: Optional[int] = None
    confidence_level: Optional[int] = None
    notes: Optional[str] = None


class WrongAnswerReviewResponse(BaseModel):
    """오답 복습 기록 응답용 모델"""
    id: int
    user_answer: str
    is_correct: bool
    time_taken: Optional[int]
    confidence_level: Optional[int]
    notes: Optional[str]
    created_at: datetime
    
    class Config:
        from_attributes = True


class WrongAnswerFilter(BaseModel):
    """오답 노트 필터링용 모델"""
    subject: Optional[str] = None
    unit: Optional[str] = None
    difficulty_min: Optional[float] = None
    difficulty_max: Optional[float] = None
    mastered: Optional[bool] = None
    source_type: Optional[str] = None
    limit: Optional[int] = 20
    offset: Optional[int] = 0


class WrongAnswerStats(BaseModel):
    """오답 노트 통계용 모델"""
    total_count: int
    mastered_count: int
    review_needed_count: int
    by_subject: dict
    by_unit: dict
    by_difficulty: dict
    recent_reviews: List[WrongAnswerReviewResponse]
